swift/ql/lib/codeql/swift/elements/expr/EnumElementExpr.qll

```diff
@@ -0,0 +1,46 @@
+private import codeql.swift.elements.expr.Expr
+private import codeql.swift.elements.expr.Argument
+private import codeql.swift.elements.expr.CallExpr
+private import codeql.swift.elements.expr.DeclRefExpr
+private import codeql.swift.elements.expr.MethodLookupExpr
+private import codeql.swift.elements.decl.EnumElementDecl
+
+/**
+ * An expression that constructs a case of an enum.
+ */
+abstract class EnumElementExpr extends Expr {
```

Comment:
We have a general rule about not having abstract AST classes since they're a terrible footgun. If a query writes a subclass that extends `EnumElementExpr` they're _extending_ the set of `EnumElementExpr` (whereas they might think they're defining a class that describes a subset of already existing `EnumElementExpr`s). This causes every cached predicate that depend on `Expr` to be re-evaluated (controlflow, dataflow, anything else we might construct in the future).

Instead, we have to define this as a non-`astract` class, and enumerate all the subclasses in `EnumElementExpr`s characteristic predicate. This ensures that users get the right behavior when they extend the class.